site/source/_posts/2017/11/17-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,263 @@
+---
+title: Dojo 2 Beta 4
+date: 2017-11-17 18:00:00
+authorId: kitsonk
+featured_image: featured.png
+---
+
+The Beta 4 Release of Dojo 2 puts us on a clear path to releasing Dojo 2.  It has several significant changes and updates.
+
+{% asset_img featured.png feature-full %}
+
+<!-- more -->
+
+## Getting Beta 4
+
+Beta 4 is currently tagged as `latest` in npm, so all just updating to the latest version will provide you with Beta 4.  We also have a release tag of `beta4` in npm if you want to focus just on using Beta 4.
+
+With Beta 3, we revised our package versions to make it easier to keep a compatible set of packes.  We are releasing all the pre-release versions of Dojo 2 packages under major version `0`, but are introducing breaking changes in the minor release (e.g. `0.2.0` has breaking changes from `0.1.0`).  The versions we publish to npm rely upon a minor range.  You will see that some packages for Beta 4 are `0.3.#` and some are `0.2.#`, but they all depend on a range of version they know they are compatible with.  This means you don't have to rely upon a release tag.  So ideally you would want to ensure if you are managing your dependencies directly, you depend upon a minor release (e.g. `~0.2.0`) for packages.
+
+As we make releases of packages in anticipation of Beta 5, we will not publish those packages to `latest` until we release Beta 5.
+
+## Integrated virtual DOM
+
+This was the biggest fundamental change for the Beta 4 release.  Instead of depending on the [Maquette](https://maquettejs.org/) virtual DOM library, we realised that our specialised requirements and needs were better solved in the medium and long term by integrating it directly.  So in Beta 4 we have replaced Maquette with `@dojo/widget-core/vdom`.
+
+This has allowed us to directly merge our abstraction of widgets and virtual DOM nodes with the virtual DOM nodes that get rendered to the DOM.  This increases the performance of the widgeting system, but also allowed us to make other performance and memory improvements to the widgeting system.
+
+This alone would not have had an effect on users of Dojo 2, but it allowed us to address a whole list of features we desired and refactoring of code that introduced some significant **breaking changes** in Beta 4.
+
+## Virtual DOM classes
+
+We felt that one of the biggest usability issues we had with Dojo 2 had to do with dealing with classes.  In Beta 4, we have changed this significantly, and while it is a **breaking change** we think the end result is _a lot_ more usable.
+
+Previously, `classes` were actually an object map of keys with a boolean value.  If you were using the theming system, you would use a function to specify your classes in a more functional way, which would generate a memoised map for you.  If you weren't you would have to do this yourself.  While it might be ok to author, it was really difficult to test.  It would look something like this:
```

Comment:
If you weren't you would have to do this yourself.  While it might be ok to author, it was really difficult to test.  It would look something like this: -> If you were not using the theming system, you would need to memoize classes directly.  While this approach might be ok to author, it was really challenging to test.  It would look something like this: